code/modules/hydroponics/beekeeping/beehive.dm

```diff
@@ -1,308 +1,455 @@
+	new/obj/effect/effect/smoke/bees(get_turf(src))
+	update_icon()
+	return TRUE
+
+/obj/structure/beehive/proc/add_frame(var/obj/item/honey_frame/H, var/mob/user)
+	if(!open)
+		to_chat(user, SPAN_NOTICE("You need to open \the [src] first before inserting \the [H]."))
+		return
+	if(LAZYLEN(frames) >= maxFrames)
+		to_chat(user, SPAN_NOTICE("There is no room for another frame."))
+		return
+
+	user.visible_message(SPAN_NOTICE("\The [user] starts placing \the [H] into \the [src]."), SPAN_NOTICE("You start placing \the [H] into of \the [src]..."))
+	if(user.do_skilled(2 SECONDS, SKILL_BOTANY, src) && user.unEquip(H, src) && !(QDELETED(src) || QDELETED(user) || QDELETED(H)))
+		LAZYADD(frames, H)
+		user.visible_message(SPAN_NOTICE("\The [user] loads \the [H] into \the [src]."), SPAN_NOTICE("You load \the [H] into \the [src]."))
+		update_icon()
+		return TRUE
+
+/obj/structure/beehive/proc/remove_frame(var/mob/user)
+	if(!LAZYLEN(frames))
+		to_chat(user, SPAN_NOTICE("There's nothing to remove from the hive. Add some frames!"))
+		return
+	if((time_end_smoked <= REALTIMEOFDAY) && bee_count)
+		to_chat(user, SPAN_WARNING("The bees won't let you take anything out like this, smoke them first."))
+		return
 
+	var/obj/item/honey_frame/H = LAZYACCESS(frames, LAZYLEN(frames))
+	user.visible_message(SPAN_NOTICE("\The [user] starts taking \the [H] out of \the [src]."), SPAN_NOTICE("You start taking \the [H] out of \the [src]..."))
+	if(user.do_skilled(3 SECONDS, SKILL_BOTANY, src) && !(QDELETED(src) || QDELETED(user) || QDELETED(H)))
+		LAZYREMOVE(frames, H)
+		user.put_in_hands(H)
+		user.visible_message(SPAN_NOTICE("\The [user] removed \the [H] out of \the [src]."), SPAN_NOTICE("You removed \the [H] out of \the [src]..."))
+		update_icon()
+		return TRUE
+
+/obj/structure/beehive/proc/apply_starter_pack(var/obj/item/bee_pack/B, var/mob/user)
+	if(!open)
+		to_chat(user, SPAN_NOTICE("You need to open \the [src] before moving the bees."))
+		return
+	if(B.full)
+		if(bee_count)
+			to_chat(user, SPAN_NOTICE("\The [src] already has bees inside."))
+			return
+		if(user.do_skilled(5 SECONDS, SKILL_BOTANY, src) && !(QDELETED(user) || QDELETED(src) || QDELETED(B)) && B.full)
+			user.visible_message(SPAN_NOTICE("\The [user] puts the queen and the bees from \the [B] into \the [src]."), SPAN_NOTICE("You put the queen and the bees from \the [B] into \the [src]."))
+			bee_count = 20
+			B.empty()
+	else 
+		if(bee_count < 90)
+			to_chat(user, SPAN_NOTICE("\The [src] is not ready to split."))
+			return
+		if((REALTIMEOFDAY > time_end_smoked))
+			to_chat(user, SPAN_NOTICE("Smoke \the [src] first!"))
+			return
+		if(user.do_skilled(5 SECONDS, SKILL_BOTANY, src) && !(QDELETED(user) || QDELETED(src) || QDELETED(B)) && !B.full && (bee_count >= 90))
+			user.visible_message(SPAN_NOTICE("\The [user] puts bees and larvae from \the [src] into \the [B]."), SPAN_NOTICE("You put bees and larvae from \the [src] into \the [B]."))
+			bee_count /= 2
```

Comment:
Needs a `round()` Nitpick also, as far as I understand `*= 0.5` is faster than `/= 2`

Reply:
I thought the other day lohikar said divisions were faster than multiplication in byond for some reasons? XD
Also, I can't round it with the code in its current state, because it the way bees "increase" they have to keep track of the whole number..